ide/api.lsp/src/org/netbeans/api/lsp/StructureElement.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.api.lsp;
+
+import java.util.List;
+import java.util.Set;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.modules.lsp.StructureElementAccessor;
+
+/**
+ * StructureElement is a tree item that shows the structure of the source code.
+ *
+ * @author Petr Pisl
+ * @since 1.8
+ */
+public final class StructureElement {
+
+    static {
+        StructureElementAccessor.setDefault(new StructureElementAccessor() {
+            @Override
+            public StructureElement createStructureElement(String name, String detail, int selectionStartOffset, int selectionEndOffset, int expandedStartOffset, int expandedEndOffset, Kind kind, Set<Tag> tags, List<StructureElement> children) {
+                return new StructureElement(name, detail, selectionStartOffset, selectionEndOffset, expandedStartOffset, expandedEndOffset, kind, tags, children);
+            }
+        });
+    }
+
+    private final String name;
+    private final String detail;
+    private final int selectionStartOffset;
```

Comment:
A weird idea: if `javax.swing.text.Position` is used here, the structure could survive some edits between the time StructureElement is created and transofmed into LSP4J wire format -- and during the translation the document could be readlocked to protect the consistency. @JaroslavTulach  ?

Reply:
Do we want the _structure_ to represent read-only snapshot and be always re-created from scratch? Or do we want the _structure_ to be updated?

I believe we strive for the first approach. As such making the fields `final` more properly represents the _snapshot_ nature of the _structure_.